### PR TITLE
Make Distant Terrain configurable

### PR DESCRIPTION
**Pull request by akortunov**
Implements [feature #4890](https://gitlab.com/OpenMW/openmw/issues/4890).

Based on [0af77e529e1c9771b7ee8e6e09df8db1959098c8](https://gitlab.com/OpenMW/openmw/merge_requests/51/diffs?commit_id=0af77e529e1c9771b7ee8e6e09df8db1959098c8), but this PR is a more clean.

As for terrain optimization, basically it is implemented [here](https://gitlab.com/OpenMW/openmw/merge_requests/51/diffs?commit_id=0bc02727b842cbbe249d9586202e8e57e4f7f3e5). Unfortunately, with this commit terrain near border disappear sometimes (bzzt managed to fix this issue somewhere in later commits).

### CODE REVIEW

**Comment by terabyte25** [comment]
I suppose #2165 needs to be changed in accordance to this if this gets merged.

**Comment by akortunov** [comment]
Ready for review now.

File: components/terrain/quadtreeworld.cpp [hunk 1]
```diff
@@ -285,7 +295,31 @@ void traverseToCell(QuadTreeNode* node, ViewData* vd, int cellX, int cellY)
...
+    int lod = Log2(int(node->getSize()));
+    if (vertexLodMod > 0)
+    {
+        lod = std::max(0, lod-vertexLodMod);
+    }
+    else if (vertexLodMod < 0)
+    {
+        float size = node->getSize();
+        // stop simplyfing at this level since smaller chunks are already less dense compared to the farther stuff in view relative terms
+        // the cutoff is 1 because at size = 1 a node has getCellVertices() (ie 65) vertices just like larger nodes
```
Comment (unelsson) [inline components/terrain/quadtreeworld.cpp]: Explanation (in comments) could be clarified altogether, but at least simplyfing -> simplifying. Thinking of better way to say "less dense compared to the farther stuff in view", but I wonder if I understand the code correctly... My personal preference wouble be to have the comment before the first "if" that is related to the comment (line 306 in this case), but not sure if that's actually better.

File: components/terrain/chunkmanager.hpp [hunk 2]
```diff
@@ -32,14 +32,17 @@ namespace Terrain
 
         osg::ref_ptr<osg::Node> getChunk(float size, const osg::Vec2f& center, int lod, unsigned int lodFlags);
 
+        void setCullingActive(bool active) { mCullingActive = active; }
+        void setCompositeMapSize(unsigned int size) { mCompositeMapSize = size; }
+        void setCompositeMapLevel(float level) { mCompositeMapLevel = level; }
+        void setMaxCompositeGeometrySize(float maxCompGeometrySize) { mMaxCompGeometrySize = maxCompGeometrySize; }
+
```
Comment (unelsson) [inline components/terrain/chunkmanager.hpp]: I haven't seen function definitions doing something in headers in OpenMW yet, is this the preferred way for such short functions?

**Comment by unelsson** [comment]
I tested this PR, played with various settings, and as far as I know, it works as intended. Reviewed with limited understanding, not much to comment really. Good job both of you.

File: components/terrain/chunkmanager.hpp (hunk 2 above)
Comment (akortunov) [reply components/terrain/chunkmanager.hpp]: Actually, we use approach (in the `apps/openmw/mwgui/formatting.hpp`), for example.
From what I can tell, there is no strict rule here.

**Comment by Capostrophic** [comment]
I think "Optimize terrain QuadTree build" commit probably doesn't belong here and moving setCullingActive doesn't change anything in particular (other than *maybe* making the implementation file more clean).

**Comment by akortunov** [comment]
> I think "Optimize terrain QuadTree build" commit probably doesn't belong here

Sorry, it was moved here by accident.

**Comment by Capostrophic** [comment]
Did some playtesting. Works nice, and the new value for 'max composite geometry size' makes gameplay with TR quite a bit smoother. Thumbs up for bzzt and akortunov.

**Pull request merged**